File: WoofAndroid/konsist/src/test/kotlin/com/perrystreet/woof/konsist/tests/ViewModelsHaveTests.kt
package com.perrystreet.woof.konsist.tests

import com.perrystreet.woof.konsist.Assertions.assertTrue
import com.perrystreet.woof.konsist.KonsistUtils
import com.perrystreet.woof.konsist.LintRuleMessage
import io.kotest.core.spec.style.BehaviorSpec

class ViewModelsHaveTests : BehaviorSpec() {
    init {
        Given("A ViewModel") {
            val viewModels = KonsistUtils.viewModels
            val testClassNames = KonsistUtils.testCode.classes().map { it.name }

            Then("A matching *ViewModelTest exists") {
                viewModels.assertTrue(message = Message) { viewModel -> "${viewModel.name}Test" in testClassNames }
            }
        }
    }

    private companion object {
        private val Message = LintRuleMessage(
            rule = "Every ViewModel has a <Name>Test.",
            why = "The ViewModel test is the single place where a feature's behaviour is specified end to end, from user action down to the fake data source.",
            howToFix = "Add <Name>Test extending ViewModelBehaviorSpec in the feature's test source set.",
            badExample = "class ProfileWoofViewModel  // no ProfileWoofViewModelTest",
            goodExample = "class ProfileWoofViewModelTest : ViewModelBehaviorSpec()",
        )
    }
}

File: WoofIOS/SwiftPackages/HarmonizeRules/Tests/HarmonizeRulesTests/Tests/ViewModelsHaveTests.swift
import Foundation
import Harmonize
import HarmonizeSemantics
import Quick

final class ViewModelsHaveTests: QuickSpec {
    override class func spec() {
        Given("A ViewModel") {
            let viewModels = WoofHarmonize.viewModels
            let testClassNames = Set(WoofHarmonize.testCode.classes().map(\.name))

            Then("A matching *ViewModelTest exists") {
                viewModels.assertTrue(rule: rule) { testClassNames.contains($0.name + "Test") }
            }
        }
    }

    private static let rule = Rule(
        description: "Every ViewModel has a <Name>Test.",
        rationale: "The ViewModel test is the single place where a feature's behaviour is specified end to end, from user action down to the fake data source.",
        fixHint: "Add <Name>Test as a QuickSpec in the feature's test target.",
        badExample: "final class ProfileWoofViewModel  // no ProfileWoofViewModelTest",
        goodExample: "final class ProfileWoofViewModelTest: QuickSpec"
    )
}
